ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/location/CaptureLocationTaskViewModel.kt

```diff
@@ -17,21 +17,71 @@ package com.google.android.ground.ui.datacollection.tasks.location
 
 import android.content.res.Resources
 import android.location.Location
+import androidx.lifecycle.viewModelScope
+import com.google.android.ground.ui.common.BaseMapViewModel
 import com.google.android.ground.ui.datacollection.tasks.AbstractTaskViewModel
 import com.google.android.ground.ui.datacollection.tasks.location.CaptureLocationTaskResult.Companion.toCaptureLocationResult
 import javax.inject.Inject
 import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+
+/** Location lock states relevant for attempting to enable it or not. */
+private enum class LocationLockEnabledState {
+  /** The default, unknown state. */
+  UNKNOWN,
+
+  /** The location lock was already enabled, or an attempt was made. */
+  ALREADY_ENABLED,
+
+  /** The location lock was not already enabled. */
+  NEEDS_ENABLE,
+
+  /** Trigger to enable the location lock. */
+  ENABLE,
+}
 
 class CaptureLocationTaskViewModel @Inject constructor(resources: Resources) :
   AbstractTaskViewModel(resources) {
 
   private val lastLocation = MutableStateFlow<CaptureLocationTaskResult?>(null)
+  /** Allows control for triggering the location lock programmatically. */
+  private val enableLocationLockFlow = MutableStateFlow(LocationLockEnabledState.UNKNOWN)
 
   suspend fun updateLocation(location: Location) {
     lastLocation.emit(location.toCaptureLocationResult())
   }
 
   fun updateResponse() {
-    setValue(lastLocation.value)
+    if (lastLocation.value == null) {
+      viewModelScope.launch { enableLocationLockFlow.emit(LocationLockEnabledState.ENABLE) }
+    } else {
+      setValue(lastLocation.value)
+    }
+  }
+
+  fun enableLocationLock() {
+    if (enableLocationLockFlow.value == LocationLockEnabledState.NEEDS_ENABLE) {
+      viewModelScope.launch { enableLocationLockFlow.emit(LocationLockEnabledState.ENABLE) }
+    }
+  }
+
+  suspend fun onMapReady(mapViewModel: BaseMapViewModel) {
+    val locationLockEnabledState =
+      if (mapViewModel.hasLocationPermission()) {
+        // User has permission to enable location updates, enable it now.
+        mapViewModel.enableLocationLockAndGetUpdates()
```

Comment:
Won't this cause the dialog to appear in the previous task too?

Reply:
I thought we were calling `enableLocationLockAndGetUpdates` only via `collect` below

Reply:
> Won't this cause the dialog to appear in the previous task too?
I tested to make sure that this is true, but the permission dialog does not show if the user already has permission, so we can preemptively start collecting location to preserve the existing behavior of front-loading the location tracking before the user arrives on the task (otherwise, there's a delay before the first data arrives).

> I thought we were calling enableLocationLockAndGetUpdates only via collect below
Meanwhile, the collect below is a fallback to turn on location tracking later manually when we need to. We could throw it in the `else` clause so that we only collect if we fall into the "no permissions" case, but I think having a noop consumer that's always there is somewhat useful. I'm happy to make the change though!

Reply:
Thanks! Marking this as resolved. Given the complexity of the method, maybe it can benefit with some refactoring (maybe in a follow-up)